File: android/src/main/kotlin/com/yourcompany/flutter_ble_core/BleUtils.kt
package com.yourcompany.flutter_ble_core

import android.bluetooth.BluetoothGattCharacteristic

object BleUtils {
    fun event(type: String, data: Map<String, Any?>): Map<String, Any?> {
        return mapOf("type" to type, "data" to data)
    }

    fun errorEvent(code: String, message: String?): Map<String, Any?> {
        return event("error", mapOf("code" to code, "message" to (message ?: "")))
    }

    fun bytesToIntList(value: ByteArray?): List<Int> {
        return value?.map { it.toInt() and 0xFF } ?: emptyList()
    }

    fun canRead(characteristic: BluetoothGattCharacteristic): Boolean {
        return characteristic.properties and BluetoothGattCharacteristic.PROPERTY_READ != 0
    }

    fun canWrite(characteristic: BluetoothGattCharacteristic): Boolean {
        val props = characteristic.properties
        return props and
            (
                BluetoothGattCharacteristic.PROPERTY_WRITE or
                    BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE
            ) != 0
    }

    fun canNotify(characteristic: BluetoothGattCharacteristic): Boolean {
        val props = characteristic.properties
        return props and
            (
                BluetoothGattCharacteristic.PROPERTY_NOTIFY or
                    BluetoothGattCharacteristic.PROPERTY_INDICATE
            ) != 0
    }
}

File: ios/flutter_ble_core/Sources/flutter_ble_core/BleUtils.swift
import CoreBluetooth
import Flutter

enum BleUtils {
    static func event(_ type: String, _ data: [String: Any?]) -> [String: Any] {
        return ["type": type, "data": data]
    }

    static func errorEvent(_ code: String, _ message: String?) -> [String: Any] {
        return event("error", ["code": code, "message": message ?? ""])
    }

    static func bytes(from data: FlutterStandardTypedData?) -> Data {
        return data?.data ?? Data()
    }

    static func canRead(_ characteristic: CBCharacteristic) -> Bool {
        return characteristic.properties.contains(.read)
    }

    static func canWrite(_ characteristic: CBCharacteristic) -> Bool {
        return characteristic.properties.contains(.write)
            || characteristic.properties.contains(.writeWithoutResponse)
    }

    static func canNotify(_ characteristic: CBCharacteristic) -> Bool {
        return characteristic.properties.contains(.notify)
            || characteristic.properties.contains(.indicate)
    }
}
